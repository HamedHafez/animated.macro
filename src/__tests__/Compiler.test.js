const Compiler = require("../Compiler");

describe("Compiler", () => {
  let parse;

  beforeAll(() => {
    const compiler = new Compiler();

    parse = (strings, ...values) => ({
      tokens: [...compiler.tokenizeTemplate(strings, ...values)],
      ast: compiler.parseTemplate(strings, ...values)
    });
  });

  test("numeric literals", () => {
    expect(parse` 1 `).toMatchSnapshot();
    expect(parse` 5. `).toMatchSnapshot();
    expect(parse` 1.2e10 `).toMatchSnapshot();
    expect(parse` 22e-1 `).toMatchSnapshot();
  });

  test("binary operators", () => {
    expect(parse` ${Math.PI}+1 `).toMatchSnapshot();
    expect(parse` ${Math.PI}-5. `).toMatchSnapshot();
    expect(parse` ${Math.PI} * 1.2e10 `).toMatchSnapshot();
    expect(parse` ${Math.PI} / 22e-1 `).toMatchSnapshot();
    expect(parse` ${Math.PI} % ${2} `).toMatchSnapshot();
  });

  test("reserved binary operators", () => {
    expect(() => parse` ${Math.PI} ** 1 `).toThrowErrorMatchingSnapshot();
  });

  test("reserved unary operators", () => {
    expect(() => parse` ++ ${Math.PI} `).toThrowErrorMatchingSnapshot();
    expect(() => parse` -- ${Math.PI} `).toThrowErrorMatchingSnapshot();
    expect(() => parse` ${Math.PI} ++ `).toThrowErrorMatchingSnapshot();
    expect(() => parse` ${Math.PI} -- `).toThrowErrorMatchingSnapshot();
  });

  test("operator precedence", () => {
    expect(parse` 1 + 2 * 3 `).toMatchSnapshot();
    expect(parse` (1 + 2) * 3 `).toMatchSnapshot();
    expect(parse` 1 - 2 / 3 `).toMatchSnapshot();
    expect(parse` (1 - 2) / 3 `).toMatchSnapshot();
    expect(parse` (1 - 2 - 3)`).toMatchSnapshot();
    expect(parse` 1 / (0 + 1) `).toMatchSnapshot();
  });

  test("binary operators", () => {
    expect(parse` ${Math.PI}+1 `).toMatchSnapshot();
    expect(parse` ${Math.PI}-5. `).toMatchSnapshot();
    expect(parse` ${Math.PI} * 1.2e10 `).toMatchSnapshot();
    expect(parse` ${Math.PI} / 22e-1 `).toMatchSnapshot();
    expect(parse` ${Math.PI} % ${2} `).toMatchSnapshot();
  });

  test("block comments", () => {
    expect(parse` /*1*/ 2 / 3 * 4 /* 5 */ `).toMatchSnapshot();
    expect(parse` /* /* 1*/0 `).toMatchSnapshot();
  });

  test("line comments", () => {
    expect(parse` 1 / 2 // 3`).toMatchSnapshot();
    expect(parse` 1 / 2 // 3\n + 4`).toMatchSnapshot();
  });
});
