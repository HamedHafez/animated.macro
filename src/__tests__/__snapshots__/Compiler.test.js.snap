// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler binary operators 1`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    "operator": "+",
    "right": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "1",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    Object {
      "type": "Operator",
      "value": "+",
    },
    Object {
      "type": "NumericLiteral",
      "value": "1",
    },
  ],
}
`;

exports[`Compiler binary operators 2`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    "operator": "-",
    "right": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "5.",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    Object {
      "type": "Operator",
      "value": "-",
    },
    Object {
      "type": "NumericLiteral",
      "value": "5.",
    },
  ],
}
`;

exports[`Compiler binary operators 3`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    "operator": "*",
    "right": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "1.2e10",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    Object {
      "type": "Operator",
      "value": "*",
    },
    Object {
      "type": "NumericLiteral",
      "value": "1.2e10",
    },
  ],
}
`;

exports[`Compiler binary operators 4`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    "operator": "/",
    "right": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "22e-1",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    Object {
      "type": "Operator",
      "value": "/",
    },
    Object {
      "type": "NumericLiteral",
      "value": "22e-1",
    },
  ],
}
`;

exports[`Compiler binary operators 5`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    "operator": "%",
    "right": Object {
      "exprType": "number",
      "index": 1,
      "type": "Placeholder",
      "value": 2,
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    Object {
      "type": "Operator",
      "value": "%",
    },
    Object {
      "index": 1,
      "type": "Placeholder",
      "value": 2,
    },
  ],
}
`;

exports[`Compiler binary operators 6`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    "operator": "+",
    "right": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "1",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    Object {
      "type": "Operator",
      "value": "+",
    },
    Object {
      "type": "NumericLiteral",
      "value": "1",
    },
  ],
}
`;

exports[`Compiler binary operators 7`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    "operator": "-",
    "right": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "5.",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    Object {
      "type": "Operator",
      "value": "-",
    },
    Object {
      "type": "NumericLiteral",
      "value": "5.",
    },
  ],
}
`;

exports[`Compiler binary operators 8`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    "operator": "*",
    "right": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "1.2e10",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    Object {
      "type": "Operator",
      "value": "*",
    },
    Object {
      "type": "NumericLiteral",
      "value": "1.2e10",
    },
  ],
}
`;

exports[`Compiler binary operators 9`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    "operator": "/",
    "right": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "22e-1",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    Object {
      "type": "Operator",
      "value": "/",
    },
    Object {
      "type": "NumericLiteral",
      "value": "22e-1",
    },
  ],
}
`;

exports[`Compiler binary operators 10`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    "operator": "%",
    "right": Object {
      "exprType": "number",
      "index": 1,
      "type": "Placeholder",
      "value": 2,
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "index": 0,
      "type": "Placeholder",
      "value": 3.141592653589793,
    },
    Object {
      "type": "Operator",
      "value": "%",
    },
    Object {
      "index": 1,
      "type": "Placeholder",
      "value": 2,
    },
  ],
}
`;

exports[`Compiler block comments 1`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "left": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "2",
      },
      "operator": "/",
      "right": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "3",
      },
      "type": "BinaryExpression",
    },
    "operator": "*",
    "right": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "4",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "type": "BlockComment",
      "value": "/*1*/",
    },
    Object {
      "type": "NumericLiteral",
      "value": "2",
    },
    Object {
      "type": "Operator",
      "value": "/",
    },
    Object {
      "type": "NumericLiteral",
      "value": "3",
    },
    Object {
      "type": "Operator",
      "value": "*",
    },
    Object {
      "type": "NumericLiteral",
      "value": "4",
    },
    Object {
      "type": "BlockComment",
      "value": "/* 5 */",
    },
  ],
}
`;

exports[`Compiler block comments 2`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "type": "NumericLiteral",
    "value": "0",
  },
  "tokens": Array [
    Object {
      "type": "BlockComment",
      "value": "/* /* 1*/",
    },
    Object {
      "type": "NumericLiteral",
      "value": "0",
    },
  ],
}
`;

exports[`Compiler line comments 1`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "1",
    },
    "operator": "/",
    "right": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "2",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "type": "NumericLiteral",
      "value": "1",
    },
    Object {
      "type": "Operator",
      "value": "/",
    },
    Object {
      "type": "NumericLiteral",
      "value": "2",
    },
    Object {
      "type": "LineComment",
      "value": "// 3",
    },
  ],
}
`;

exports[`Compiler line comments 2`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "left": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "1",
      },
      "operator": "/",
      "right": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "2",
      },
      "type": "BinaryExpression",
    },
    "operator": "+",
    "right": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "4",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "type": "NumericLiteral",
      "value": "1",
    },
    Object {
      "type": "Operator",
      "value": "/",
    },
    Object {
      "type": "NumericLiteral",
      "value": "2",
    },
    Object {
      "type": "LineComment",
      "value": "// 3",
    },
    Object {
      "type": "Operator",
      "value": "+",
    },
    Object {
      "type": "NumericLiteral",
      "value": "4",
    },
  ],
}
`;

exports[`Compiler numeric literals 1`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "type": "NumericLiteral",
    "value": "1",
  },
  "tokens": Array [
    Object {
      "type": "NumericLiteral",
      "value": "1",
    },
  ],
}
`;

exports[`Compiler numeric literals 2`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "type": "NumericLiteral",
    "value": "5.",
  },
  "tokens": Array [
    Object {
      "type": "NumericLiteral",
      "value": "5.",
    },
  ],
}
`;

exports[`Compiler numeric literals 3`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "type": "NumericLiteral",
    "value": "1.2e10",
  },
  "tokens": Array [
    Object {
      "type": "NumericLiteral",
      "value": "1.2e10",
    },
  ],
}
`;

exports[`Compiler numeric literals 4`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "type": "NumericLiteral",
    "value": "22e-1",
  },
  "tokens": Array [
    Object {
      "type": "NumericLiteral",
      "value": "22e-1",
    },
  ],
}
`;

exports[`Compiler operator precedence 1`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "1",
    },
    "operator": "+",
    "right": Object {
      "exprType": "Animated.Value",
      "left": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "2",
      },
      "operator": "*",
      "right": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "3",
      },
      "type": "BinaryExpression",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "type": "NumericLiteral",
      "value": "1",
    },
    Object {
      "type": "Operator",
      "value": "+",
    },
    Object {
      "type": "NumericLiteral",
      "value": "2",
    },
    Object {
      "type": "Operator",
      "value": "*",
    },
    Object {
      "type": "NumericLiteral",
      "value": "3",
    },
  ],
}
`;

exports[`Compiler operator precedence 2`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "left": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "1",
      },
      "operator": "+",
      "right": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "2",
      },
      "type": "BinaryExpression",
    },
    "operator": "*",
    "right": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "3",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "type": "LeftParen",
      "value": "(",
    },
    Object {
      "type": "NumericLiteral",
      "value": "1",
    },
    Object {
      "type": "Operator",
      "value": "+",
    },
    Object {
      "type": "NumericLiteral",
      "value": "2",
    },
    Object {
      "type": "RightParen",
      "value": ")",
    },
    Object {
      "type": "Operator",
      "value": "*",
    },
    Object {
      "type": "NumericLiteral",
      "value": "3",
    },
  ],
}
`;

exports[`Compiler operator precedence 3`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "1",
    },
    "operator": "-",
    "right": Object {
      "exprType": "Animated.Value",
      "left": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "2",
      },
      "operator": "/",
      "right": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "3",
      },
      "type": "BinaryExpression",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "type": "NumericLiteral",
      "value": "1",
    },
    Object {
      "type": "Operator",
      "value": "-",
    },
    Object {
      "type": "NumericLiteral",
      "value": "2",
    },
    Object {
      "type": "Operator",
      "value": "/",
    },
    Object {
      "type": "NumericLiteral",
      "value": "3",
    },
  ],
}
`;

exports[`Compiler operator precedence 4`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "left": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "1",
      },
      "operator": "-",
      "right": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "2",
      },
      "type": "BinaryExpression",
    },
    "operator": "/",
    "right": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "3",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "type": "LeftParen",
      "value": "(",
    },
    Object {
      "type": "NumericLiteral",
      "value": "1",
    },
    Object {
      "type": "Operator",
      "value": "-",
    },
    Object {
      "type": "NumericLiteral",
      "value": "2",
    },
    Object {
      "type": "RightParen",
      "value": ")",
    },
    Object {
      "type": "Operator",
      "value": "/",
    },
    Object {
      "type": "NumericLiteral",
      "value": "3",
    },
  ],
}
`;

exports[`Compiler operator precedence 5`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "left": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "1",
      },
      "operator": "-",
      "right": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "2",
      },
      "type": "BinaryExpression",
    },
    "operator": "-",
    "right": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "3",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "type": "LeftParen",
      "value": "(",
    },
    Object {
      "type": "NumericLiteral",
      "value": "1",
    },
    Object {
      "type": "Operator",
      "value": "-",
    },
    Object {
      "type": "NumericLiteral",
      "value": "2",
    },
    Object {
      "type": "Operator",
      "value": "-",
    },
    Object {
      "type": "NumericLiteral",
      "value": "3",
    },
    Object {
      "type": "RightParen",
      "value": ")",
    },
  ],
}
`;

exports[`Compiler operator precedence 6`] = `
Object {
  "ast": Object {
    "exprType": "Animated.Value",
    "left": Object {
      "exprType": "Animated.Value",
      "type": "NumericLiteral",
      "value": "1",
    },
    "operator": "/",
    "right": Object {
      "exprType": "Animated.Value",
      "left": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "0",
      },
      "operator": "+",
      "right": Object {
        "exprType": "Animated.Value",
        "type": "NumericLiteral",
        "value": "1",
      },
      "type": "BinaryExpression",
    },
    "type": "BinaryExpression",
  },
  "tokens": Array [
    Object {
      "type": "NumericLiteral",
      "value": "1",
    },
    Object {
      "type": "Operator",
      "value": "/",
    },
    Object {
      "type": "LeftParen",
      "value": "(",
    },
    Object {
      "type": "NumericLiteral",
      "value": "0",
    },
    Object {
      "type": "Operator",
      "value": "+",
    },
    Object {
      "type": "NumericLiteral",
      "value": "1",
    },
    Object {
      "type": "RightParen",
      "value": ")",
    },
  ],
}
`;

exports[`Compiler reserved binary operators 1`] = `"The operator '**' is not supported in this context"`;

exports[`Compiler reserved unary operators 1`] = `"The operator '++' is not supported in this context"`;

exports[`Compiler reserved unary operators 2`] = `"The operator '--' is not supported in this context"`;

exports[`Compiler reserved unary operators 3`] = `"The operator '++' is not supported in this context"`;

exports[`Compiler reserved unary operators 4`] = `"The operator '--' is not supported in this context"`;

exports[`Compiler unsupported syntax 1`] = `
"invalid syntax near:

   Math.PI 
   ^"
`;

exports[`Compiler unsupported syntax 2`] = `
"invalid syntax near:

   abs(
   ^"
`;

exports[`Compiler unsupported syntax 3`] = `
"invalid syntax near:

   \\"Hi\\" 
   ^"
`;

exports[`Compiler unsupported syntax 4`] = `
"invalid syntax near:

   if (
   ^"
`;

exports[`Compiler unsupported syntax 5`] = `
"invalid syntax near:

   true 
   ^"
`;

exports[`Compiler unsupported syntax 6`] = `
"invalid syntax near:

  .interpolate() 
   ^"
`;
